rfcs/text/0030-linux-event-model.md

```diff
@@ -0,0 +1,1121 @@
+          type: 'char_device',
+          char_device: {
+            major: 8,
+            minor: 1
+          }
+        }
+      ],
+      tty: {
+        descriptor: 0,
+        type: 'char_device',
+        char_device: {
+          major: 8,
+          minor: 1
+        }
+      }
+    },
+    fds: [
+      {
+        descriptor:0,
+        type: 'char_device',
+        char_device: {
+          major: 8,
+          minor: 1
+        }
+      },
+      {
+        descriptor:1,
+        type:'pipe',
+        pipe: {
+          inode: '6183207'
+        }
+      }
+    ],
+    tty: {
+      descriptor: 0,
+      type: 'char_device',
+      char_device: {
+        major: 8,
+        minor: 1
+      }
+    }
+  }
+}
+```
+
+<!--1
+Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting, or if on the larger side, add them to the corresponding RFC folder.
+-->
+
+<!--
+Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
+-->
+
+## Scope of impact
+
+<!--
+Stage 2: Identifies scope of impact of changes. Are breaking changes required? Should deprecation strategies be adopted? Will significant refactoring be involved? Break the impact down into:
+ * Ingestion mechanisms (e.g. beats/logstash)
+ * Usage mechanisms (e.g. Kibana applications, detections)
+ * ECS project (e.g. docs, tooling)
+The goal here is to research and understand the impact of these changes on users in the community and development teams across Elastic. 2-5 sentences each.
+-->
+
+## Concerns
+
+<!--
+Stage 1: Identify potential concerns, implementation challenges, or complexity. Spend some time on this. Play devil's advocate. Try to identify the sort of non-obvious challenges that tend to surface later. The goal here is to surface risks early, allow everyone the time to work through them, and ultimately document resolution for posterity's sake.
+-->
+
+Potential concerns for me are primary tied to message size. The nested ancestors and their many fields could add up. That being said, there is a larger concern wrt number of events that will dwarf any issue around message size. That effort will come down to defining a set of event filters to eliminate unwanted noise. Scraping /proc and sending up all process executions can be very load heavy task. This issue is probably outside of the scope of ECS though.
+
+Another point of concern is the re-use of process.entity_id. There is nuance around how this ID is calculated, and we are proposing a new hash/calculation based of a different set of fields. This issue is being tracked here: https://github.com/elastic/security-team/issues/2458
```

Comment:
Same thing as below, can this info be brought into a public issue or otherwise summarized here so everyone can get the context?

Reply:
The entity_id note should likely be removed; syntactically the entity_id will remain the same.  @ferullo has the semantic impact of making this identifier follow the process from its creation to its exit (vs changing upon execs) seemingly covered.

I will summarize the size concerns/trade-offs here shortly.

Reply:
## Why is this additional process context worth the extra bytes/bandwidth/storage?

Each ECS process event originally had information about the process itself and its parent process. This RFC adds full process information for the following related processes to each process event.
* the process's session leader
  * this provides a lot of information about the execution context of a process
  * For example, if the session leader is a webserver like nginx, process execution other than nginx can be flagged as anomalous
  * if it's a shell (e.g. bash) with an associated controlling terminal (i.e. likely driven by a human) extra scrutiny can be applied
* the process's entry-session leader
  * this identifies how the host was initially accessed so one could disallow/alert on sensitive activities via SSH but allow them via IPMI/serial.
  * this is sometimes the same as the session leader
* the process's process group leader
  * this is part of enabling pipelines like "cat foo | grep bar | sort | less" to be reassembled for rendering or search/detection.

This RFC also adds the following *anemic* versions of the process object (i.e. those with just entity_id, pid and starttime) to allow chained queries for forensics and to help detect user-entered commands vs automatically spawned child processes.
* the session leader's parent process
  * allows navigating upward through sessions by querying for the parent process and its session id
* the parent process's group leader
  * allows checking if your parent's process group is different from the process's. If so, this process is likely a user-entered command

These additions allow:

* More precise alerting with less noise
  * For example, instead of alerting on any use of sudo or nmap, alert only if the entry session type is SSH from a non-private IP address
  * this information is in each event; multi-stage queries are not required
  * query languages such as KQL can consider data that was only possible with EQL "sequence" queries
  * EQL sequence queries will miss information that happened outside of the alert rule's event loading window
* Improved readability (session view rendering) vs log lines style views
  * Session view stitches together information-dense, collapsable views of the session's hierarchical process model
  * The extra context allows re-assembly much of the session tree without loading every event in a session
  * sessions can run for months (tmux) and may have events that have been deleted (such as the original session leader/bash) that will be filled in by this extra context.
* Improved search efficiency and expressiveness
  * the extra context allows for efficient, very targeted queries for forensics; what would have taken several queries can be done it one query
  * (forensics and finding events for session views)

## Is this the most compact representation we could use?

No, we have optimized for queriability at the cost of compactness. A single process may have several roles:
* it could be both the session leader and the entry session
* it could be both main process and group leader

In these cases, it would be more space efficient to tag the process with its roles (e.g. primary process and group leader) instead of repeating it with a different field name.   However, reduced queriability significantly reduces the value of this information.


## If data volume becomes an issue how can it be mitigated?

Rather than removing information from every process event, one can remove/filter entire process trees.  The filtered process events can still generate alerts as with existing Endpoint EventFilters.

A classic example of noise people do not want to store is Kubernetes Pod health checks.  These executions can be detected and filtered at the Endpoint.

Reply:
@mitodrummer I think the part about scraping proc can be removed - implementation detail and it's an infrequent occurrence (at Endpoint startup). The bulk of these events will be sourced from BPF with most of their extra context information added without touching /proc.

Reply:
@m-sample added your text in place of the above, and revised some of the other sections I wrote. I will mark this resolved. Feel free to re-open or create a new comment if there is anything else. cc @epixa